pkg/intermediate/aggregate.go

```diff
@@ -124,74 +140,272 @@ func (a *AggregationProcess) DeleteFlowKeyFromMap(flowKey FlowKey) {
 	delete(a.flowKeyRecordMap, flowKey)
 }
 
-// correlateRecords fills records info by correlating incoming and current records
-func (a *AggregationProcess) correlateRecords(flowKey FlowKey, record entities.Record) {
+// addOrUpdateRecordInMap either adds the record to flowKeyMap or updates the record in
+// flowKeyMap by doing correlation or updating the stats.
+func (a *AggregationProcess) addOrUpdateRecordInMap(flowKey *FlowKey, record entities.Record) error {
 	a.mutex.Lock()
 	defer a.mutex.Unlock()
-	existingRecords := a.flowKeyRecordMap[flowKey]
-	// only fill the information for record from source node
-	if isRecordFromSrc(record) {
-		var isFilled bool
-		for _, existingRec := range existingRecords {
-			for _, field := range a.correlateFields {
-				if ieWithValue, exist := record.GetInfoElementWithValue(field); exist {
-					existingIeWithValue, _ := existingRec.GetInfoElementWithValue(field)
-					ieWithValue.Value = existingIeWithValue.Value
-					isFilled = true
-				}
+
+	aggregationRecord, exist := a.flowKeyRecordMap[*flowKey]
+	if exist {
+		if !isRecordIntraNode(record) {
+			// Do correlation of records if record belongs to inter-node flow and
+			// records from source and destination node are not received.
+			if !aggregationRecord.ReadyToSend && !areRecordsFromSameNode(record, aggregationRecord.Record) {
+				a.correlateRecords(record, aggregationRecord.Record)
+				aggregationRecord.ReadyToSend = true
 			}
-			if isFilled {
-				break
+			// Aggregation of incoming flow record with existing by updating stats
+			// and flow timestamps.
+			if isRecordFromSrc(record) {
+				a.aggregateRecords(record, aggregationRecord.Record, true, false)
+			} else {
+				a.aggregateRecords(record, aggregationRecord.Record, false, true)
 			}
+		} else {
+			// For intra-node flows, just do aggregation of the flow record with
+			// existing record by updating the stats and flow timestamps. Correlation
+			// is not required.
+			a.aggregateRecords(record, aggregationRecord.Record, true, true)
 		}
 	} else {
-		for _, existingRec := range existingRecords {
-			if isRecordFromSrc(existingRec) {
-				for _, field := range a.correlateFields {
-					if ieWithValue, exist := record.GetInfoElementWithValue(field); exist {
-						existingIeWithValue, _ := existingRec.GetInfoElementWithValue(field)
-						existingIeWithValue.Value = ieWithValue.Value
+		// Add all the new stat fields and initialize them.
+		if !isRecordIntraNode(record) {
+			if isRecordFromSrc(record) {
+				if err := a.addFieldsForStatsAggregation(record, true, false); err != nil {
+					return err
+				}
+			} else {
+				if err := a.addFieldsForStatsAggregation(record, false, true); err != nil {
+					return err
+				}
+			}
+		} else {
+			if err := a.addFieldsForStatsAggregation(record, true, true); err != nil {
+				return err
+			}
+		}
+		aggregationRecord = AggregationFlowRecord{
+			record,
+			false,
+			true,
+		}
+		if isRecordIntraNode(record) {
```

Comment:
This seems not consistent with line 151, where ReadyToSend is set to true when !isRecordIntraNode(record) ?